### PR TITLE
Country based telemetry synchronization

### LINKED ISSUE #177: China telemetry in iOS
## Description
Implement telemetry collection and sending from China.
iOS part of https://github.com/mapbox/mapbox-vision/issues/1312.

## Goal
Support collection and uploading telemetry from China in iOS SDK.

## Steps
See https://github.com/mapbox/mapbox-vision/issues/1312.

### PR DESCRIPTION
**Pull request by chezzdev**
Checks:

* [x] Update changelog
* [x] Rebase to `dev` branch
* [x] Assign reviewers

Linked issues:
Resolves #177

### CODE REVIEW

File: MapboxVision/Services/Recording/RecordCoordinator.swift [hunk 1]
```diff
@@ -60,50 +59,55 @@ final class RecordCoordinator {
...
-        videoRecorder.chunkLimit = savesSourceVideo ? 1 : defaultChunkLimit
-        videoRecorder.startRecording(to: cachePath, settings: videoSettings)
+            self.currentStartTime = DispatchTime.now()
+            self.currentEndTime = nil
 
-        delegate?.recordingStarted(path: recordingPath.recordingPath)
-    }
+            self.videoRecorder.chunkLength = self.savesSourceVideo ? 0 : defaultChunkLength
+            self.videoRecorder.chunkLimit = self.savesSourceVideo ? 1 : defaultChunkLimit
+            self.videoRecorder.startRecording(to: cachePath, settings: videoSettings)
```
[the remark this answers is not in the document]
Comment (chezzdev) [reply MapboxVision/Services/Recording/RecordCoordinator.swift]: 👌

File: MapboxVision/Services/Sync/RecordSynchronizer.swift [hunk 2]
```diff
@@ -7,36 +7,54 @@ private let updatingInterval = 1 * .hour
...
         let archiver: Archiver
         let fileManager: FileManagerProtocol
     }
 
+    private enum State {
+        case idle
+        case syncing
+        case stopping
+
+        var isIdle: Bool {
```
[the remark this answers is not in the document]
Comment (chezzdev) [reply MapboxVision/Services/Sync/RecordSynchronizer.swift]: No problem, I'll remove it.

File: MapboxVision/Services/SyncRegion.swift [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import Foundation
+
+enum SyncRegion: String {
+    case china = "China"
+    case other = "Other"
```
[the remark this answers is not in the document]
Comment (chezzdev) [reply MapboxVision/Services/SyncRegion.swift]: There's no `default` region. There's "China" and others.
At the same time, there is a `default` endpoint in the events SDK and this is its representation.

File: MapboxVisionTests/Mocks/MockSynchronizable.swift [hunk 4]
```diff
@@ -2,9 +2,48 @@ import Foundation
...
 
 class MockSynchronizable: Synchronizable {
+    enum Action: Equatable {
+        case sync
+        case stopSync
+        case set(dataSource: RecordDataSource, baseURL: URL?)
+
+         static func == (lhs: Action, rhs: Action) -> Bool {
+            switch (lhs, rhs) {
+            case (.sync, .sync):
```
[the remark this answers is not in the document]
Comment (chezzdev) [reply MapboxVisionTests/Mocks/MockSynchronizable.swift]: Ok.

File: MapboxVisionTests/VisionManagerTests.swift [hunk 5]
```diff
@@ -75,34 +101,40 @@ class VisionManagerTests: XCTestCase {
...
         // Then
-        // we expect that visionManager will call SessionRecorder's method `startInternal` for all countries but China and that there
-        // will no be any other calls to SessionRecorder. For China we expect `stop` method to be called
+        // VisionManager is expected to start internal recording on its start, not react to country update
+        // if synchronization region isn't changed, stop internal recording and start it again if the synchronization region is changed
         let expectedActions: [MockSessionRecorder.Action] = [
-            .startInternal, // USA
-            .startInternal, // UK
+            .startInternal, // initial
+            // nothing      // USA
```
[the remark this answers is not in the document]
Comment (chezzdev) [reply MapboxVisionTests/VisionManagerTests.swift]: Well, these are hints to the reader that changing a country inside one region shouldn't initiate any action. Do you think they're misleading or something?

File: MapboxVision/Services/Constants.swift [hunk 6]
```diff
@@ -6,4 +6,10 @@ enum Constants {
     static let millisecondsInSecond: Double = 1000
 
     static let frameDuration: TimeInterval = 1 / 30.0
+
+    enum URL {
+        static let defaultEventsEndpoint = "https://events.mapbox.com"
+
```
[the remark this answers is not in the document]
Comment (chezzdev) [reply MapboxVision/Services/Constants.swift]: Ok.

File: MapboxVisionTests/BaseVisionManagerTests.swift [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        visionManager = BaseVisionManager(dependencies: dependencies)
+    }
+
+    func testChangingCountries() {
+        // Given
+        let actions: [(Country, [MockSynchronizable.Action])] = [
+            (.unknown, [.stopSync]),
+            (.USA, [
+                .stopSync,
+                .set(dataSource: otherDataSource, baseURL: URL(string: Constants.URL.defaultEventsEndpoint)!),
```
[the remark this answers is not in the document]
Comment (chezzdev) [reply MapboxVisionTests/BaseVisionManagerTests.swift]: Not sure what do you mean. I guess this line tests exactly that.

File: MapboxVisionTests/BaseVisionManagerTests.swift [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+        synchronizer = MockSynchronizable()
+        dependencies = BaseDependencies(native: MockNative(), synchronizer: synchronizer)
+        visionManager = BaseVisionManager(dependencies: dependencies)
+    }
+
+    func testChangingCountries() {
+        // Given
+        let actions: [(Country, [MockSynchronizable.Action])] = [
+            (.unknown, [.stopSync]),
```
[the remark this answers is not in the document]
Comment (chezzdev) [reply MapboxVisionTests/BaseVisionManagerTests.swift]: What would you suggest?

File: MapboxVision/Services/Recording/DocumentsLocation.swift [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        if
+            let searchDirectory = searchDirectory,
+            let baseDirectory = NSSearchPathForDirectoriesInDomains(searchDirectory, .userDomainMask, true).first
+        {
+            basePath = baseDirectory
+        } else {
+            basePath = NSTemporaryDirectory()
+        }
+
+        if let bundleIdentifier = Bundle(for: BundleToken.self).bundleIdentifier {
```
[the remark this answers is not in the document]
Comment (dersim-davaod) [reply MapboxVision/Services/Recording/DocumentsLocation.swift]: Oh now I see, thanks :)

In that case I suggest to mark this class with `final`.
Also do we have explicit `init()` for `BundleToken`. Would default initializer be enough?

File: MapboxVision/Services/Recording/DocumentsLocation.swift [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+        case .custom:
+            searchDirectory = nil
+        }
+
+        var basePath: String
+
+        if
+            let searchDirectory = searchDirectory,
+            let baseDirectory = NSSearchPathForDirectoriesInDomains(searchDirectory, .userDomainMask, true).first
+        {
```
Comment (dersim-davaod) [inline MapboxVision/Services/Recording/DocumentsLocation.swift]: Probably we can simplify `path` implementation a bit then?

```swift
    var path: String {
        var basePath: String

        switch self {
        case .currentRecording, .cache, .recordings:
            if let cachesDirPath = NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true).first {
                basePath = cachesDirPath
            } else {
                basePath = NSTemporaryDirectory()
            }
        case .custom:
            basePath = NSTemporaryDirectory()
        }

        if let bundleIdentifier = Bundle(for: BundleToken.self).bundleIdentifier {
            basePath = basePath.appendingPathComponent(bundleIdentifier, isDirectory: true)
        }

        return basePath.appendingPathComponent(directoryName, isDirectory: true)
    }
```

File: MapboxVision/Services/Recording/DocumentsLocation.swift [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+            searchDirectory = .applicationSupportDirectory
+        default:
+            searchDirectory = nil
+        }
+
+        var basePath: String
+
+        if
+            let searchDirectory = searchDirectory,
+            let baseDirectory = NSSearchPathForDirectoriesInDomains(searchDirectory, .userDomainMask, true).first
```
[the remark this answers is not in the document]
Comment (dersim-davaod) [reply MapboxVision/Services/Recording/DocumentsLocation.swift]: Decided to reimplement this somewhen later. Corresponding issue is: https://github.com/mapbox/mapbox-vision-ios/issues/194

File: MapboxVision/Services/SyncRegion.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (dersim-davaod) [reply MapboxVision/Services/SyncRegion.swift]: Agreed to leave is as it is.

File: MapboxVision/Services/Recording/DocumentsLocation.swift (hunk 9 above)
[the remark this answers is not in the document]
Comment (chezzdev) [reply MapboxVision/Services/Recording/DocumentsLocation.swift]: Will add `final`. Private `init` is there to restrict `BundleIdentifier` instantiating.

File: MapboxVision/VisionManager/VisionManager.swift [hunk 12]
```diff
@@ -266,18 +299,19 @@ extension VisionManager: VideoSourceObserver {
 extension VisionManager: RecordCoordinatorDelegate {
     func recordingStarted(path: String) {}
 
-    func recordingStopped() {
+    func recordingStopped(recordingPath: RecordingPath) {
+        try? handle(recordingPath: recordingPath, country: currentCountry)
```
Comment (IodaMikeMapbox) [inline MapboxVision/VisionManager/VisionManager.swift]: Here we should handle not current country, but the country which was known before the moment of stopping, we can assume that if before stopping there was country `.unknown` it is equal to the new one

File: MapboxVision/Services/Recording/DocumentsLocation.swift (hunk 10 above)
Comment (chezzdev) [reply MapboxVision/Services/Recording/DocumentsLocation.swift]:
```swift
    var path: String {
        var basePath: String = NSTemporaryDirectory()

        if
            self != .custom,
            let cachesPath = NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true).first
        {
            basePath = cachesPath
        }

        if let bundleIdentifier = Bundle(for: BundleToken.self).bundleIdentifier {
            basePath = basePath.appendingPathComponent(bundleIdentifier, isDirectory: true)
        }

        return basePath.appendingPathComponent(directoryName, isDirectory: true)
    }
```
?

Comment (dersim-davaod) [reply MapboxVision/Services/Recording/DocumentsLocation.swift]: Looks good!

**Pull request merged**